### DISCUSSION TITLE
CsvHelper.ReaderException: System.OutOfMemoryException

### DISCUSSION THREAD

**Discussion by VenkatReddySR** [JoshClose/CsvHelper#1519] [Q&A]
Am reading around 200k records 
It throws an exception (CsvHelper.ReaderException: An unexpected error occurred. ---> System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.Dynamic.ExpandoObject.ExpandoData.UpdateClass(ExpandoClass newClass))
at below line of code     
                
var record = csv.GetRecord<dynamic>() as IDictionary<string, object>;

If am doing wrong please specify
Below is the code

```
using (var reader = new StringReader(fileString))
            using (var csv = new CsvReader(reader))
            {
                csv.Configuration.Delimiter = "\t";
                csv.Configuration.MissingFieldFound = null;
                csv.Read();
                csv.ReadHeader();
                var headerRow = csv.Context.HeaderRecord;

                while (csv.Read())
                {
                    var record = csv.GetRecord<dynamic>() as IDictionary<string, object>;
                    csvRecords.Add(record);
                }
            }
```

**Comment by JoshClose** [comment]
If you have a `StringReader` that means the entire file is in memory. You're then looping the records and putting the entire files into memory again when adding to `csvRecords`.

To conserve memory you could use a `Stream` and a `StreamReader` so that only a buffered amount is in memory. I'll need to know what you're doing with the records to give a suggestion for that, but ideally you would only have a single record in memory at once.

**Comment by VenkatReddySR** [comment]
Dear Josh,

Thanks for response, Am writing output records to a file with few modifications

**Comment by VenkatReddySR** [comment]
Dear Josh,

As per your suggestion , I have removed **StringReader** and used StreamReader, Yet I have noticed out of memory exception, below is the code 

```
           using (var reader = new StreamReader(GenerateStreamFromString(fileString)))
            using (var csv = new CsvReader(reader))
            {
                csv.Configuration.Delimiter = "\t";
                csv.Configuration.MissingFieldFound = null;
                csv.Read();
                csv.ReadHeader();
                while (csv.Read())
                {
//**Exception comes at below line**
         IDictionary<string, object> record = csv.GetRecord<dynamic>() as IDictionary<string, object>;
          csvRecords.Add(dictRecord);
                }
            }  

   private static Stream GenerateStreamFromString(string fileContent)
        {
            var stream = new MemoryStream();
            var writer = new StreamWriter(stream);
            writer.Write(fileContent);
            writer.Flush();
            stream.Position = 0;
            return stream;
        }
```
